From 3c6fb5d9b521d09cc73eb4bf99c0240e87d7a160 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Mon, 15 Aug 2022 21:33:37 +0300
Subject: [PATCH] [SW-100596] pcie: gaudi3: add snps wa for dcc bug

Recently SNPS found a bug in their PCIe PHY. This patch adds
SW workaround for DCC snps bug as part of pcie link up sequence in
ROM recovery flow.

Change-Id: I06b9d982874b4fdbb4ca3d0198c2c012a70e99f3
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index dacb1f5..534759f 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -340,6 +340,24 @@
 	regmap_writel(D0_PCIE_AUX_BOOT_TERM_CLR_CLR_M, mmD0_PCIE_AUX_BOOT_TERM_CLR);
 }
 
+#define PHY_CR_PARA_CMD_DATA		0x2
+#define PHY_CR_PARA_CMD_ADDR		0x5149
+
+void pcie_wa_snps_dcc(void)
+{
+	uint32_t val;
+
+	regmap_writel(D0_PCIE_PHY_CR_PARA_CFG_SEL_M, mmD0_PCIE_PHY_CR_PARA_CFG);
+
+	regmap_writel(D0_PCIE_PHY_CR_PARA_CTRL_WR_EN_M, mmD0_PCIE_PHY_CR_PARA_CTRL);
+	val = ((PHY_CR_PARA_CMD_DATA << D0_PCIE_PHY_CR_PARA_CMD_DATA_S) | PHY_CR_PARA_CMD_ADDR);
+	regmap_writel(val, mmD0_PCIE_PHY_CR_PARA_CMD);
+
+	/* Wait until status is cleared */
+	while (regmap_readl(mmD0_PCIE_PHY_CR_PARA_STS))
+		;
+}
+
 static void arcboot_rom_prstn_init(void)
 {
 	regmap_writel(0, mmD0_PSOC_GLOBAL_CONF_PRSTN_MASK);
diff --git a/common/pcie.c b/common/pcie.c
index d20ec52..f637da9 100644
--- a/common/pcie.c
+++ b/common/pcie.c
@@ -75,6 +75,7 @@
 __weak void pcie_config_boot_termination(void) {}
 __weak void pcie_phy_set_lane_refclk_mode(void) {}
 __weak void pcie_phy_set_pll_short_lock(void) {}
+__weak void pcie_wa_snps_dcc(void) {}
 
 void pcie_config(void)
 {
@@ -113,6 +114,8 @@
 
 		pcie_wait_for_sram_init_done();
 
+		pcie_wa_snps_dcc();
+
 		pcie_phy_set_sram_ext_ld_done();
 
 		pcie_set_pcie_device_vendor_id();
